compressors/tANS.py

```diff
@@ -0,0 +1,380 @@
+"""tANS v1 (table ANS) implementation 
+
+NOTE: tANS v1 is ina  way cached rANS implementation. There are other variants of tANS possible
+See the wiki link: https://github.com/kedartatwawadi/stanford_compression_library/wiki/Asymmetric-Numeral-Systems
+for more details on the algorithm
+
+## References
+1. Original Asymmetric Numeral Systems paper:  https://arxiv.org/abs/0902.0271
+2. https://github.com/kedartatwawadi/stanford_compression_library/wiki/Asymmetric-Numeral-Systems
+More references in the wiki article
+"""
+
+from dataclasses import dataclass
+import numpy as np
+from typing import Tuple, Any, List
+from core.data_encoder_decoder import DataDecoder, DataEncoder
+from utils.bitarray_utils import BitArray, get_bit_width, uint_to_bitarray, bitarray_to_uint
+from core.data_block import DataBlock
+from core.prob_dist import Frequencies, get_mean_log_prob
+from utils.test_utils import get_random_data_block, try_lossless_compression
+from utils.misc_utils import cache, is_power_of_two
+import pprint
+from compressors.rANS import rANSParams, rANSEncoder, rANSDecoder
+
+
+@dataclass
+class tANSParams(rANSParams):
+    """
+    NOTE: params are same as rANSParams
+    we just restrict some parameter values
+    """
+
+    def __post_init__(self):
+        super().__post_init__()
+        ## restrict some param values for tANS
+        # to make the rANS cachable, the total_freq needs to be a power of 2
+        assert is_power_of_two(
+            self.freqs.total_freq
+        ), "Please normalize self.freqs.total_freq to be a power of two"
+
+        # NOTE: NUM_BITS_OUT != 1, probably doesn't make practical sense for tANS?
```

Comment:
Can you elaborate why a similar technique as rANS can't work for tANS for NUM_BITS_OUT>0?
Also, what's the typical practice (e.g., in FSE library)?

Reply:
Sorry, I don't understand this. Why would it be exponential in `NUM_BITS_OUT`? The table size shouldn't depend on how many bits I am sending to ensure the range remains in [L,H]. It should depend on the parameters once the state is in that range. Am I missing something?

Reply:
`[L, H] = [M, 2^{NUM_OUT_BITS}*M - 1]` .. due to this range increases exponentially with NUM_OUT_BITS

Reply:
but ll add a comment

Reply:
Thanks!

Reply:
done